File: android/app/src/main/java/com/flatlistpro/AutoLayoutShadow.kt
package com.flatlistpro

class AutoLayoutShadow {
    var horizontal: Boolean = false
    var scrollOffset: Int = 0
    var windowSize: Int = 0
    var renderOffset = 0

    /** Checks for overlaps or gaps between adjacent items and then applies a correction.
     * Performance: RecyclerListView renders very small number of views and this is not going to trigger multiple layouts on Android side. Not expecting any major perf issue. */
    fun clearGapsAndOverlaps(sortedItems: Array<CellContainer>) {
        var currentMax = 0

        for (i in 0 until sortedItems.size - 1) {
            val cell = sortedItems[i]
            val neighbour = sortedItems[i + 1]
            if (isWithinBounds(cell)) {
                if (!horizontal) {
                    currentMax = kotlin.math.max(currentMax, cell.bottom);
                    if (cell.left < neighbour.left) {
                        if (cell.right != neighbour.left) {
                            neighbour.right = cell.right + neighbour.width
                            neighbour.left = cell.right
                        }
                        if (cell.top != neighbour.top) {
                            neighbour.bottom = cell.top + neighbour.height
                            neighbour.top = cell.top
                        }
                    } else {
                        neighbour.bottom = currentMax + neighbour.height
                        neighbour.top = currentMax

                    }
                } else {
                    currentMax = kotlin.math.max(currentMax, cell.right);
                    if (cell.top < neighbour.top) {
                        if (cell.bottom != neighbour.top) {
                            neighbour.bottom = cell.bottom + neighbour.height
                            neighbour.top = cell.bottom
                        }
                        if (cell.left != neighbour.left) {
                            neighbour.right = cell.left + neighbour.width
                            neighbour.left = cell.left
                        }
                    } else {
                        neighbour.right = currentMax + neighbour.width
                        neighbour.left = currentMax
                    }
                }
            }
        }
    }

    /** It's important to avoid correcting views outside the render window. An item that isn't being recycled might still remain in the view tree. If views outside get considered then gaps between
     * unused items will cause algorithm to fail.*/
    private fun isWithinBounds(cell: CellContainer): Boolean {
        return if (!horizontal) {
            (cell.top >= (scrollOffset - renderOffset) || cell.bottom >= (scrollOffset - renderOffset)) &&
                    (cell.top <= scrollOffset + windowSize || cell.bottom <= scrollOffset + windowSize)
        } else {
            (cell.left >= (scrollOffset - renderOffset) || cell.right >= (scrollOffset - renderOffset)) &&
                    (cell.left <= scrollOffset + windowSize || cell.right <= scrollOffset + windowSize)
        }
    }
}

File: harmony/flash_list/src/main/cpp/AutoLayoutShadow.cpp
//
// Created on 2024/3/3
//
// Node APIs are not fully supported. To solve the compilation error of the interface cannot befoun,
// please include "napai/native_api.h".

#include <arkui/native_interface.h>
#include <react/renderer/components/view/ViewProps.h>
#include <glog/logging.h>
#include <chrono>
#include <sys/param.h>
#include <thread>
#include "AutoLayoutShadow.h"

namespace rnoh {

    /**checks for overlaps or gaps.between adjacent items and then applies a correction (Only Grid layouts with varying
     * spans) Performance: RecyclerListView renders very small number of views and this is not going to trigger multiple
     * layouts on Android side. Not expecting any major perf issu.*/
    void AutoLayoutShadow::clearGapsAndOverlaps(std::vector<ComponentInstance::Shared> sortedItems) {
        if (sortedItems.empty()) {
            return;
        }
        int maxBound = 0;
        int minBound = INT_MAX;
        int maxBoundNeighbour = 0;
        lastMaxBoundOverall = 0;
        for (int i = 0; i < sortedItems.size() - 1; i++) {
            auto cell = std::dynamic_pointer_cast<rnoh::CellContainerComponentInstance>(sortedItems[i]);
            auto neighbour = std::dynamic_pointer_cast<rnoh::CellContainerComponentInstance>(sortedItems[i + 1]);
            // Only apply correction if the next cell is consecutive.
            bool isNeighbourConsecutive = neighbour->getIndex() == cell->getIndex() + 1;
            if (isWithinBounds(*cell)) {
                if (!horizontal) {
                    maxBound = MAX(maxBound, cell->getBottom());
                    minBound = MIN(minBound, cell->getTop());
                    maxBoundNeighbour = maxBound;
                    if (isNeighbourConsecutive) {
                        if (cell->getLeft() < neighbour->getLeft()) {
                            if (cell->getRight() != neighbour->getLeft()) {
                                neighbour->setRight(cell->getRight() + neighbour->getWidth());
                            }
                            if (cell->getTop() != neighbour->getTop()) {
                                neighbour->setBottom(cell->getTop() + neighbour->getHeight());
                                neighbour->setTop(cell->getTop());
                            }
                        } else {
                            neighbour->setBottom(maxBound + neighbour->getHeight());
                            neighbour->setTop(cell->getTop());
                        }
                    }
                    if (isWithinBounds(*neighbour)) {
                        maxBoundNeighbour = MAX(maxBound, neighbour->getBottom());
                    }
                } else {
                    maxBound = MAX(maxBound, cell->getRight());
                    minBound = MIN(minBound, cell->getLeft());
                    maxBoundNeighbour = maxBound;
                    if (isNeighbourConsecutive) {
                        if (cell->getTop() < neighbour->getTop()) {
                            if (cell->getBottom() != neighbour->getTop()) {
                                neighbour->setBottom(cell->getBottom() + neighbour->getHeight());
                                neighbour->setTop(cell->getBottom());
                            }
                            if (cell->getLeft() != neighbour->getLeft()) {
                                neighbour->setRight(cell->getLeft() + neighbour->getWidth());
                                neighbour->setLeft(cell->getLeft());
                            }
                        } else {
                            neighbour->setRight(maxBound + neighbour->getWidth());
                            neighbour->setLeft(maxBound);
                        }
                    }
                    if (isWithinBounds(*neighbour)) {
                        maxBoundNeighbour = MAX(maxBound, neighbour->getRight());
                    }
                }
            }
            if (horizontal) {
                lastMaxBoundOverall = MAX(lastMaxBoundOverall, cell->getRight());
                lastMaxBoundOverall = MAX(lastMaxBoundOverall, neighbour->getRight());
            } else {
                lastMaxBoundOverall = MAX(lastMaxBoundOverall, cell->getBottom());
                lastMaxBoundOverall = MAX(lastMaxBoundOverall, neighbour->getBottom());
            }
            cell->setLayout(cell->getLayoutMetrics());
            neighbour->setLayout(neighbour->getLayoutMetrics());
        }
        lastMaxBound = maxBoundNeighbour;
        lastMinBound = minBound;
    }

    /** Offset provider by react can be one frame behind the real one, it's important that this method is called with
     * offset taken directly from scrollview object*/
    int AutoLayoutShadow::computeBlankFromGivenOffset(int actualScrollOffset, int distanceFromWindowStart,
                                                      int distanceFromWindowEnd) {
        auto scrollOffset = actualScrollOffset - offsetFromStart;
        blankOffsetAtStart = lastMinBound - scrollOffset - distanceFromWindowStart;
        blankOffsetAtEnd =
            scrollOffset + windowSize - renderOffset - lastMaxBound - distanceFromWindowEnd;
        return MAX(blankOffsetAtStart, blankOffsetAtEnd);
    }

    /**It's importance to aviod correcting views outside the render window. An item that isn't being recycled might
     * still remain in the view tree. If views outside get considered then gaps between unused items will cause
     * algorithm to fail.*/
    bool AutoLayoutShadow::isWithinBounds(CellContainerComponentInstance &cell) {
        auto scrollOffset = this->scrollOffset - offsetFromStart;
        if (!this->horizontal) {
            return (cell.getTop() >= (scrollOffset - renderOffset) ||
                    cell.getBottom() >= (scrollOffset - renderOffset)) &&
                   (cell.getTop() <= scrollOffset + windowSize ||
                    cell.getBottom() <= scrollOffset + windowSize);
        } else {
            return (cell.getLeft() >= (scrollOffset - renderOffset) ||
                    cell.getRight() >= (scrollOffset - renderOffset)) &&
                   (cell.getLeft() <= scrollOffset + windowSize ||
                    cell.getRight() <= scrollOffset + windowSize);
        }
    }
} // namespace rnoh